037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,251 @@
+# Automatic rebalance on cluster scaling
+
+This feature is about enabling the cluster operator to get an optimization proposal, which is auto-approved, in order to run the corresponding rebalancing when the user scales the cluster up or down by changing the `spec.kafka.replicas` field on the `Kafka` custom resource.
+Such a rebalancing can happen right after the scale up to move some replicas to the newly added brokers or right before the scale down to move replicas off of the brokers to be removed.
+
+## Current situation
+
+Currently, there is no automatic rebalancing when the user scales the cluster up or down.
+It means that:
+
+* after scaling up the cluster, in order to rebalance the cluster and move some replicas to the newly added brokers, the user has to create a `KafkaRebalance` custom resource using the `spec.mode: add-brokers` mode and add the new broker IDs to the the `spec.brokers` field.
+* before scaling down the cluster, in order to rebalance the cluster and move replicas off of the brokers to be removed, the user has to create a `KafkaRebalance` custom resource using the `spec.mode: remove-brokers` mode and add the broker IDs of the brokers to be removed to the `spec.brokers` field.
+
+For more details, the proposal [035](https://github.com/strimzi/proposals/blob/main/035-rebalance-types-scaling-brokers.md) introduced the rebalancing modes mentioned above.
+
+## Motivation
+
+The rebalancing operation during scaling up and down is a manual two step process.
+
+* scaling up the cluster, then running a rebalance.
+* running a rebalance, then scaling down the cluster.
+
+In order to simplify the process, it would be great for the user to deal with just scaling the cluster by changing the `spec.kafka.replicas` field and then having the operator taking care of running a rebalancing operation automatically if requested.
+
+Taking into account the `StrimziPodSets` usage (instead of the `StatefulSets`), which will be the default in the future, a more general use case will be about adding new brokers and/or removing existing ones, not talking just about scaling up and down anymore.
+
+This proposal takes into account this future plan by referring to more general operations like add and remove brokers instead of scaling up and down.
+
+## Proposal
+
+A new `KafkaRebalanceTemplate` custom resource could be a solution allowing the user to define the configuration for the auto-rebalancing but leaving to the operator the creation of the corresponding `KafkaRebalance` custom resource.
```

Comment:
Have you thought about using a regular `KafkaRebalance` as the template? I'm still a bit concerned by the efforts related to yet another CRD.

Reply:
We might be able to use the existing `mode` for this, e.g. `mode: template`, (or possibly including the existing mode in the template using `mode: add-brokers-template`, but this feels a bit ugly; I think it's cleaner when the add-brokers vs remove-brokers is not part of the template itself, that we you could use the same temple for both the add and remove cases).

Reply:
> Have you thought about using a regular KafkaRebalance as the template?

Absolutely. It was my first thought when I started to think about this. I had in mind a similar model as Tom but using a new additional field in the `KafkaRebalance`, something like a boolean `spec.template` (which was `false` by default).
It was in my mind the way to avoid the rebalance operator starting to process the `KafkaRebalance` resource when it was created (as opposed to what happens today) because just a reference to use when the auto-rebalancing was needed. But as Tom pointed it was not possible to share the same "template" across two different modes for example which is something that can be useful imho.

Reply:
I still think new CRD is a huge overhead for what it does - not just initially but also maintenance wise. But if you think you need it, fine with me I guess.